File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/R/DRarrayFn.R
DRarrayFn <- function(nXbasisw, nWbasisw, nXbasisx, nWbasisx, 
                     Bvecx, Btens) {
  DRvec <- rep(0,nXbasisw*nXbasisx*nWbasisw)
  Btens <- as.vector(Btens)
  inprodDList <- .C("DRarrayFn", as.integer(nXbasisw), 
                     as.integer(nWbasisw), 
                     as.integer(nXbasisx), 
                     as.integer(nWbasisx), 
                     as.double(Bvecx), 
                     as.double(Btens), 
                     as.double(DRvec))
  DRarrayC <- array(inprodDList[[7]],c(nXbasisw,nXbasisx,nWbasisw))
  return(DRarrayC)
}

File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/src/DRarrayFn.c
//
//  DRarrayFn.c
//  
//
//  Created by James Ramsay on 2018-3-6.
//
//

#include <stdio.h>

void DRarrayFn(int   *nXbasisw, int    *nWbasisw, int    *nXbasisx, int *nWbasisx,
              double *Bvecx,    double *Btenswx,  double *DRarray)
{
    
    int nXw = *nXbasisw;
    int nWw = *nWbasisw;
    int nXx = *nXbasisx;
    int nWx = *nWbasisx;
    
    int nDRarray = nXw*nXx*nWw;
    
    /* assign 0 to all positions */
    
    for (int p = 0; p < nDRarray; p++)
    {
        DRarray[p]=0;
    }
    
    /* loop through all columns of all four input matrices */
    
    for (int i = 0; i < nXw; i++)
    {
        for (int j = 0; j < nWw; j++)
        {
            for (int k = 0; k < nXx; k++)
            {
                for (int l = 0; l < nWx; l++)
                {
                    int ijkl = i*nWx*nXx*nWw + j*nWx*nXx + k*nWx + l;
                    
                    /*  increment value at index by the product */
                    
                    DRarray[i+k*nXw+j*nXw*nXx] += Bvecx[l] * Btenswx[ijkl];
                    
                }
                
            }
        }
    }
    
}
